weave/integrations/google_genai/gemini_utils.py

```diff
@@ -0,0 +1,145 @@
+from weave.trace.op import _add_accumulator
+from weave.trace.serialization.serialize import dictify
+from weave.trace.weave_client import Call
+
+if TYPE_CHECKING:
+    from google.genai.types import GenerateContentResponse
+
+
+def google_genai_gemini_postprocess_inputs(inputs: dict[str, Any]) -> dict[str, Any]:
+    """
+    Postprocess inputs of the trace for the Google GenAI Gemini API to be used in
+    the trace visualization in the Weave UI. If the parameter `self` is present
+    (i.e, if the function being traced is a stateful method), it is converted to a
+    dictionary of attributes that can be displayed in the Weave UI.
+    """
+    # Extract the model name from the inputs and ensure it is present in the inputs
+    model_name = getattr(inputs["self"], "_model", None)
+    if model_name is not None:
+        inputs["model"] = model_name
+
+    # Convert the `self` parameter which is actually the state of the
+    # `google.genai.models.Models` object to a dictionary of attributes that can
+    # be displayed in the Weave UI
+    if "self" in inputs:
+        inputs["self"] = dictify(inputs["self"])
+    return inputs
+
+
+def google_genai_gemini_on_finish(
+    call: Call, output: Any, exception: Optional[BaseException] = None
+) -> None:
+    """
+    On finish handler for the Google GenAI Gemini API integration that ensures the usage
+    metadata is added to the summary of the trace.
+    """
+    if not (model_name := call.inputs.get("model")):
+        raise ValueError("Unknown model type")
+    usage = {model_name: {"requests": 1}}
+    summary_update = {"usage": usage}
+    if output:
+        call.output = dictify(output)
+        if hasattr(output, "usage_metadata"):
+            usage[model_name].update(
+                {
+                    "prompt_tokens": output.usage_metadata.prompt_token_count,
+                    "completion_tokens": output.usage_metadata.candidates_token_count,
+                    "total_tokens": output.usage_metadata.total_token_count,
+                }
+            )
+    if call.summary is not None:
+        call.summary.update(summary_update)
+
+
+def google_genai_gemini_accumulator(
+    acc: Optional["GenerateContentResponse"], value: "GenerateContentResponse"
+) -> "GenerateContentResponse":
+    if acc is None:
+        return value
+
+    for i, value_candidate in enumerate(value.candidates):
+        if i >= len(acc.candidates):
+            break
+        for j, value_part in enumerate(value_candidate.content.parts):
+            if j >= len(acc.candidates[i].content.parts):
+                break
+            if value_part.text is not None:
+                acc.candidates[i].content.parts[j].text += value_part.text
+
+    if acc.usage_metadata.prompt_token_count is None:
+        acc.usage_metadata.prompt_token_count = 0
+    elif value.usage_metadata.prompt_token_count is not None:
+        acc.usage_metadata.prompt_token_count += value.usage_metadata.prompt_token_count
```

Comment:
Are the `elif` here correct?  Won't this skip the token count for the very first values?